From d613b2f77c988e2c82ca1d4dfaaaa6d7e7245116 Mon Sep 17 00:00:00 2001
From: jakub <jakub@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Tue, 30 Sep 2014 10:39:03 +0000
Subject: [PATCH 271/326] 	PR inline-asm/63282 	* ifcvt.c
 (dead_or_predicable): Don't call redirect_jump_1 	or invert_jump_1 if
 jump isn't any_condjump_p.

	* gcc.c-torture/compile/pr63282.c: New test.


git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/branches/gcc-4_8-branch@215714 138bc75d-0d04-0410-961f-82ee72b054a4
---
 gcc/ChangeLog                                 |  6 ++++++
 gcc/ifcvt.c                                   |  3 +++
 gcc/testsuite/ChangeLog                       |  5 +++++
 gcc/testsuite/gcc.c-torture/compile/pr63282.c | 13 +++++++++++++
 4 files changed, 27 insertions(+)
 create mode 100644 gcc/testsuite/gcc.c-torture/compile/pr63282.c

diff --git a/gcc/ChangeLog b/gcc/ChangeLog
index 86cc14f..71a9fbe 100644
--- a/gcc/ChangeLog
+++ b/gcc/ChangeLog
@@ -1,3 +1,9 @@
+2014-09-30  Jakub Jelinek  <jakub@redhat.com>
+
+	PR inline-asm/63282
+	* ifcvt.c (dead_or_predicable): Don't call redirect_jump_1
+	or invert_jump_1 if jump isn't any_condjump_p.
+
 2014-09-29  Charles Baylis  <charles.baylis@linaro.org>
 
 	Backport from mainline r212303
diff --git a/gcc/ifcvt.c b/gcc/ifcvt.c
index dd6df5b..70a84d2 100644
--- a/gcc/ifcvt.c
+++ b/gcc/ifcvt.c
@@ -4271,6 +4271,9 @@ dead_or_predicable (basic_block test_bb, basic_block merge_bb,
   old_dest = JUMP_LABEL (jump);
   if (other_bb != new_dest)
     {
+      if (!any_condjump_p (jump))
+	goto cancel;
+
       if (JUMP_P (BB_END (dest_edge->src)))
 	new_dest_label = JUMP_LABEL (BB_END (dest_edge->src));
       else if (new_dest == EXIT_BLOCK_PTR)
diff --git a/gcc/testsuite/ChangeLog b/gcc/testsuite/ChangeLog
index 7143edb..7bcdf7a 100644
--- a/gcc/testsuite/ChangeLog
+++ b/gcc/testsuite/ChangeLog
@@ -1,3 +1,8 @@
+2014-09-30  Jakub Jelinek  <jakub@redhat.com>
+
+	PR inline-asm/63282
+	* gcc.c-torture/compile/pr63282.c: New test.
+
 2014-09-26  Jakub Jelinek  <jakub@redhat.com>
 
 	* g++.dg/compat/struct-layout-1_generate.c: Add -Wno-abi
diff --git a/gcc/testsuite/gcc.c-torture/compile/pr63282.c b/gcc/testsuite/gcc.c-torture/compile/pr63282.c
new file mode 100644
index 0000000..cb23278
--- /dev/null
+++ b/gcc/testsuite/gcc.c-torture/compile/pr63282.c
@@ -0,0 +1,13 @@
+/* PR inline-asm/63282 */
+
+void bar (void);
+
+void
+foo (void)
+{
+  asm volatile goto ("" : : : : a, b);
+a:
+  bar ();
+b:
+  return;
+}
-- 
2.0.3

